lib/client/ipam_test.go

```diff
@@ -170,8 +180,86 @@ var _ = Describe("IPAM tests", func() {
+			if expError != nil {
+				Expect(outError).To(HaveOccurred())
+				Expect(outError.Error()).To(Equal(expError.Error()))
+			}
+		},
+
+		// Test cases (ClaimAffinity):
+		// Test 1: claim affinity for an unclaimed IPNet of size 64 - expect 1 claimed blocks, 0 failed and expect no error.
+		Entry("Claim affinity for an unclaimed IPNet of size 64", "192.168.1.0/26", "host-A", true, []string{"192.168.1.0/24", "fd80:24e2:f998:72d6::/120"}, net.IP{}, 1, 0, nil),
+
+		// Test 2: claim affinity for an unclaimed IPNet of size smaller than 64 - expect 0 claimed blocks, 0 failed and expect an error error.
+		Entry("Claim affinity for an unclaimed IPNet of size smaller than 64", "192.168.1.0/27", "host-A", true, []string{"192.168.1.0/24", "fd80:24e2:f998:72d6::/120"}, net.IP{}, 0, 0, errors.New("The requested CIDR (192.168.1.0/27) is smaller than the minimum.")),
+
+		// Test 3: claim affinity for a IPNet that has an IP already assigned to another host.
+		// - Assign an IP with AssignIP to "host-A" from a configured pool - expect 0 claimed blocks, 0 failed and expect no error.
+		Entry("Claim affinity for a IPNet that has an IP already assigned to another host (Assign IP)", "", "host-A", true, []string{"10.0.0.0/24", "fd80:24e2:f998:72d6::/120"}, net.ParseIP("10.0.0.1"), 0, 0, nil),
+
+		// - Claim affinity for "Host-B" to the block that IP belongs to - expect 3 claimed blocks and 1 failed.
+		Entry("Claim affinity for a IPNet that has an IP already assigned to another host (Claim affinity for Host-B)", "10.0.0.0/24", "host-B", false, []string{"10.0.0.0/24", "fd80:24e2:f998:72d6::/120"}, net.IP{}, 3, 1, nil),
+
+		// Test 4: claim affinity to a block twice from different hosts.
+		// - Claim affinity to an unclaimed block for "Host-A" - expect 4 claimed blocks, 0 failed and expect no error.
+		Entry("Claim affinity to an unclaimed block for Host-A", "10.0.0.0/24", "host-A", true, []string{"10.0.0.0/24", "fd80:24e2:f998:72d6::/120"}, net.IP{}, 4, 0, nil),
+
+		// - Claim affinity to the same block again but for "host-B" this time - expect 0 claimed blocks, 4 failed and expect no error.
+		Entry("Claim affinity to the same block again but for Host-B this time", "10.0.0.0/24", "host-B", false, []string{"10.0.0.0/24", "fd80:24e2:f998:72d6::/120"}, net.IP{}, 0, 4, nil),
+	)
 })
 
+// testIPAMClaimAffinity thakes inNet which is the CIDR to claim affinity to. It also takes host, poolSubnets to configure a new pool,
+// when assignIP is not empty it will assign that IP and return zero-value for the return values, and cleanEnv is to clear the datastore etc.
+func testIPAMClaimAffinity(inNet net.IPNet, host string, poolSubnet []string, assignIP net.IP, cleanEnv bool) (claimed []cnet.IPNet, failed []cnet.IPNet, outErr error) {
+	if cleanEnv {
+		testutils.CleanEtcd()
+		c, _ := testutils.NewClient("")
+		for _, v := range poolSubnet {
+			testutils.CreateNewPool(*c, v, false, false, true)
+		}
+	}
+	ic := setupIPAMClient(cleanEnv)
+
+	if len(assignIP) != 0 {
```

Comment:
I agree this is rather hacky, but it is for the test case where one IP is assigned from a block to host-A and then host-B tries to claim all the blocks, and AssignIP needs to happen after creating the client and before claimingAffinity.

What I can do is make a utility function that gets called instead of having this AssignIP login inside claimAffinity function. 
Does that sound ok or you have something else in mind?